Add a new api object MultiClusterServiceDNSRecord and a controller

shashidharatd opened the pull request:
This is with reference to the idea discussed in https://github.com/kubernetes-sigs/federation-v2/issues/56
Initiated this PR to workout on API semantics.
Request your review mainly on the API type in https://github.com/kubernetes-sigs/federation-v2/commit/baab43bc166f9b8eb7fd495575784d1551d9b869

/cc @pmorie @marun @font @irfanurrehman

pmorie commented on config/apiserver.yaml:
```diff
@@ -0,0 +1,189 @@
+    spec:
+      containers:
+      - name: apiserver
+        image: shashidharatd/fv2:v5
```
Should this stuff be a separate PR?

shashidharatd replied:
Sorry, my bad. This was not supposed to be checked in. Removed now.

shashidharatd commented:
This is now ready for review. PTAL. Thanks !

shashidharatd commented:
The test is failing because of the flake which is fixed in https://github.com/kubernetes-sigs/federation-v2/pull/79, otherwise the pr is good for review. @pmorie @marun Can you PTAL this PR. Thanks !

shashidharatd commented:
> I think it would be more valuable to implement some form of integration and e2e tests instead of unit tests with mock code for the DNS controller, because I think it will provide better overall testing coverage long-term. It can also work with the already existing framework. I do realize that testing a true load balancer status update is really only feasible in e2e with a provider that has a supported load balancer implementation in k8s. Perhaps for those cases we will have to rely on a more advanced e2e regression test that we could run periodically or as part of CI. I believe in k/federation this was done with Google's Cloud but required Google's intervention.

@font, initially i gave a thought to implement e2e/ integration tests, but as you said the current CI environment does not have clusters with provider. IMHO, this needs to be tackled as a separate task. However i would add the e2e test cases particular to this PR and keep them in pending state and we can enable them whenever we have CI with clusters having provider and these testcases can be tested manually with `minikube` setup having `metallb` controller. Please suggest any other better approaches if you have. Thanks !

font commented:
@shashidharatd What if we still implement integration and e2e tests for this feature and have a fake update to the service objects to simulate a load balancer status update? That is, I'd prefer to fake out tests in integration and e2e instead of relying on mock unit tests. Thoughts?

shashidharatd commented:
> @shashidharatd What if we still implement integration and e2e tests for this feature and have a fake update to the service objects to simulate a load balancer status update? That is, I'd prefer to fake out tests in integration and e2e instead of relying on mock unit tests. Thoughts?

@font, sorry for the delay in responding. I took your suggestion and did faking and added integration tests. IMHO, the e2e tests should not do any faking and i plan to add the real e2e tests for this feature, when we have e2e tests running in cloud environment in CI. Let me know your opinion.

font commented on pkg/controller/dns/controller_test.go:
```diff
@@ -0,0 +1,300 @@
+/*
```
I'm not sure how valuable it is to keep these unit tests around now that the integration tests are providing more complete test coverage.

shashidharatd replied:
ok, i will remove

shashidharatd commented on test/integration/integration_test.go:
```diff
@@ -28,4 +28,5 @@ func TestIntegration(t *testing.T) {
 	defer FedFixture.TearDown(tl)
 
 	t.Run("TestCrud", TestCrud)
+	t.Run("TestMultiCLusterDNS", TestMultiCLusterDNS)
```
will fix.

shashidharatd commented on test/integration/crud_test.go:
```diff
@@ -32,16 +32,12 @@ import (
 
 // TestCrud validates create/read/update/delete operations for federated types.
-func TestCrud(t *testing.T) {
+var TestCrud = func(t *testing.T) {
```
yes agree, that change is required for tests to pass. whichever lands in first, the other needs a rebase.

shashidharatd commented on test/integration/multiclusterdns_test.go:
```diff
@@ -0,0 +1,295 @@
+		lbsuffix++
+		suffix := strings.TrimPrefix(clusterName, "test-cluster-")
+		regionZones := fedv1a1.FederatedClusterStatus{
+			Region: fmt.Sprintf("region%s", suffix),
```
ok

shashidharatd commented on test/integration/multiclusterdns_test.go:
```diff
@@ -0,0 +1,295 @@
+	clusterClients        map[string]clientset.Interface
+	clusterLbs            map[string]string
+	clusterRegionZones    map[string]fedv1a1.FederatedClusterStatus
+	fedDNSStatusWithoutLb []dnsv1a1.ClusterDNS
```
This is mainly to fix golint error.
```
#golint test/integration/multiclusterdns_test.go 
test/integration/multiclusterdns_test.go:140:2: struct field fedDnsStatusWithoutLb should be fedDNSStatusWithoutLb
```

shashidharatd commented on test/integration/multiclusterdns_test.go:
```diff
@@ -0,0 +1,295 @@
+}
+
+func newDNSObject(name, namespace string) *dnsv1a1.MultiClusterDNSLb {
+	return &dnsv1a1.MultiClusterDNSLb{
```
ok. will update

shashidharatd commented on pkg/apis/multiclusterdns/v1alpha1/multiclusterdnslb_types.go:
```diff
@@ -0,0 +1,69 @@
+// MultiClusterDNSLb
+// +k8s:openapi-gen=true
+// +resource:path=multiclusterdnslbs,strategy=MultiClusterDNSLbStrategy
+type MultiClusterDNSLb struct {
```
For lack of better naming and the errors in auto-generation in apiserver-builder, i named the object like it is. Definetely its not a good name for api object.
I am inclined to name it `ServiceDNSElement`. I could name api object ending with 's' and 'a', as there are issues in auto-generation code. @pmorie @marun @irfanurrehman, request your suggestion too on this.

shashidharatd commented:
@font, i added the integration and e2e tests. can you PTAL. Thanks !

font commented on test/e2e/servicedns.go:
```diff
@@ -0,0 +1,140 @@
+/*
```
(no changes necessary) I was not suggesting we create duplicate tests in E2E that are covered in integration. It was only that we could have a common place to host the DNS tests that can be used by integration and E2E similar to the common crud tester. I think this is okay for now, but ideally we could have common test code that checks the DNS object updates regardless if we are testing a fake load balancer for integration or a real load balancer for E2E. Also, then when we consolidate integration and managed E2E in the future, we can just remove the code path for one of the implementations.

marun commented on test/e2e/framework/framework.go:
```diff
@@ -42,6 +42,7 @@ type FederationFramework interface {
 	CrClient(userAgent string) crclientset.Interface
 
 	ClusterClients(apiResource *metav1.APIResource, userAgent string) map[string]common.TestCluster
+	ClusterTypedClients(userAgent string) map[string]kubeclientset.Interface
```
How about `ClusterDynamicClients` and `ClusterKubeClients`?

font replied:
+1

shashidharatd commented on pkg/apis/multiclusterdns/v1alpha1/servicednselement_types.go:
```diff
@@ -0,0 +1,69 @@
+// ServiceDNSElement
+// +k8s:openapi-gen=true
+// +resource:path=servicednselements,strategy=ServiceDNSElementStrategy
+type ServiceDNSElement struct {
```
Finally named it MultiClusterServiceDNSRecord. Hope its fine.

shashidharatd commented on test/e2e/servicedns.go:
```diff
@@ -0,0 +1,140 @@
+	})
+
+	Describe("object", func() {
+		It("should be created, read, updated and deleted successfully", func() {
```
Updated the code to check successful deletion.

shashidharatd commented:
@font, i have addressed all the comments. PTAL. Thanks !

marun commented on pkg/controller/federatedcluster/clusterclient.go:
```diff
@@ -101,3 +109,52 @@ func (self *ClusterClient) GetClusterHealthStatus() *fedv1a1.FederatedClusterSta
 }
+
+// GetClusterZones gets the kubernetes cluster zones and region by inspecting labels on nodes in the cluster.
+func (self *ClusterClient) GetClusterZones() (zones []string, region string, err error) {
```
(No action required) Why is zones a slice if only a single value is ever returned?

marun commented on pkg/apis/federatedscheduling/v1alpha1/zz_generated.conversion.go:
```diff
@@ -1,7 +1,7 @@
 
 /*
-Copyright 2018 The Federation v2 Authors.
+Copyright 2018 The Kubernetes Authors.
```
@shashidharatd It appears this code needs to be regenerated to reflect this change to the boilerplate made when the repo was moved.  I've placed a `do-not-merge` label accordingly.

@pmorie Were you going to add a boilerplate check?

font replied:
@marun The new boilerplate specifies `The Kubernetes Authors` and is correct now that we are part of the k8s org. I asked @pmorie in https://github.com/kubernetes-sigs/federation-v2/pull/77#issuecomment-391050970 to verify because I had the same concern. That PR also added a boilerplate checker that runs as part of CI.

marun commented:
I'm guessing the switch to MultiCluster prefix was in response to the discussion in the WG meeting yesterday about ensuring unique names?  I guess it's a fair choice, given that the capability could be used separate from Federation.

What is the intended relationship between a `MultiClusterServiceDNSRecord` and a `FederatedService` when the 2 features are deployed together?  As proposed, there is no relation and a user would have to create and manage the 2 resources manually.  There is also no provision for targeting member clusters selectively.  I'm not suggesting that this question should block merge, but I would expect it be answered before alpha release if the goal is v1 parity.

I'm also a bit confused at the duplication of testing effort.  I won't block merge for this reason, but I would like to see removal of duplication going in future PRs.

---

Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 90 of 90 files at r3, 2 of 2 files at r4, 3 of 6 files at r5, 11 of 11 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/apis/multiclusterdns/v1alpha1/multiclusterservicednsrecord_types.go, line 63 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQaL7vBZ-7Nupyrtrd:-LEQaL7vBZ-7Nupyrtre:bynnn24) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/pkg/apis/multiclusterdns/v1alpha1/multiclusterservicednsrecord_types.go#L63)):*
> ```Go
> type ClusterDNS struct {
> 	// Cluster name
> 	Cluster string
> ```

Why is there no json tag for this field?

---

*[pkg/controller/servicedns/controller.go, line 369 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQcwsm5RBXxy4N2SmN:-LEQcwsm5RBXxy4N2SmO:b-8n1won) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/pkg/controller/servicedns/controller.go#L369)):*
> ```Go
> 		// and traffic is not routable to the service.
> 		// We avoid such service shards while writing DNS records.
> 		endpoints, err := c.getServiceEndpointsInCluster(cluster.Name, key)
> ```

Given that `endpoints` is only used to determine whether a service is backed by pods, why is it necessary to collect the addresses?

---

*[test/common/servicedns.go, line 64 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQfIJEDCp6i57RUIt4:-LEQfIJEDCp6i57RUIt5:bskoigx) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/common/servicedns.go#L64)):*
> ```Go
> }
> 
> func newObject(filename, name, namespace string, object pkgruntime.Object) error {
> ```

(No action required) I'm curious why you chose to implement this from scratch rather than extending the functions in testobjects.go.

I'm also not sure why you're loading fixture from disk.  This was done for the push reconciler because the rewrite to the dynamic client made it easier to define test objects as yaml, because the equivalent definition with Unstructured are very verbose.  It also allows for user-defined primitives to be validated without writing code.  In the case of the `servicedns` controller, there is no need a generic approach, and defining fixture in code seems much simpler.

---

*[test/common/servicedns.go, line 90 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQeC36EJUapr9bm2tf:-LEQeC36EJUapr9bm2tg:biv33q8) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/common/servicedns.go#L90)):*
> ```Go
> 
> // WaitForDNSObject waits for MultiClusterServiceDNSRecord object to match the desired status.
> func WaitForDNSObject(tl TestLogger, fedClient fedclientset.Interface, namespace string, desired *dnsv1a1.MultiClusterServiceDNSRecord, interval time.Duration) {
> ```

(No action required) This and the wait-for-deletion function seem like an awful lot of work for a single type.  There are already similar functions in crudtester.go that could be extended to accept functions for accessing and comparing a resource so that the type/client-specific behavior can be passed in rather than hard-coded.

---

*[test/common/servicedns.go, line 125 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQeWPF14AU-_wPaQyC:-LEQeWPF14AU-_wPaQyD:b-qy3vww) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/common/servicedns.go#L125)):*
> ```Go
> }
> 
> func equivalent(actual, desired *dnsv1a1.MultiClusterServiceDNSRecord) bool {
> ```

What about meta equivalency?

---

*[test/e2e/servicedns.go, line 62 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQl6E-DqM2qtEIu8pz:-LEQl6E-DqM2qtEIu8q-:b-7ck8kl) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L62)):*
> ```Go
> 
> 	Describe("object", func() {
> 		It("should be created, read, updated and deleted successfully", func() {
> ```

Consider revising this test name to be more descriptive.  The test as written seems to validate that the servicedns controller is correctly maintaining state without a service or endpoints.

---

*[test/e2e/servicedns.go, line 69 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQmHXhDmjGfVn9f1L3:-LEQmHXhDmjGfVn9f1L4:bvz92dj) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L69)):*
> ```Go
> 			framework.ExpectNoError(err, "Error creating MultiClusterServiceDNS object %v", serviceDNS)
> 			defer func() {
> 				fedClient.MulticlusterdnsV1alpha1().MultiClusterServiceDNSRecords(namespace).Delete(serviceDNS.GetName(), nil)
> ```

Why is it desirable to explicitly delete the resource, given that the namespace and its contents will be removed on framework teardown?

---

*[test/e2e/servicedns.go, line 70 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQlFYuCGdJCmpQfdKT:-LEQlFYuCGdJCmpQfdKU:b-isawsn) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L70)):*
> ```Go
> 			defer func() {
> 				fedClient.MulticlusterdnsV1alpha1().MultiClusterServiceDNSRecords(namespace).Delete(serviceDNS.GetName(), nil)
> 				common.WaitForDNSObjectDeletion(tl, fedClient, serviceDNS.Name, namespace, framework.PollInterval)
> ```

Why does the wait function accept a polling interval but not a timeout?  I think both should vary between integration and e2e as per the example of the crud tester.

---

*[test/e2e/servicedns.go, line 90 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQlHfA9jhO_mJqnIF4:-LEQlHfA9jhO_mJqnIF5:b-g4g94m) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L90)):*
> ```Go
> 
> 			serviceDNS.Status = serviceDNSStatus
> 			common.WaitForDNSObject(tl, fedClient, namespace, serviceDNS, framework.PollInterval)
> ```

Ditto the comment about varying the timeout.

---

*[test/e2e/servicedns.go, line 92 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQm3oX9Cd3aCvWfEmH:-LEQm-TIEW841q0J431H:b-53b1bv) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L92)):*
> ```Go
> 			common.WaitForDNSObject(tl, fedClient, namespace, serviceDNS, framework.PollInterval)
> 
> 			serviceDNS.Spec.FederationName = "myfed"
> ```

What is being checked in the remainder of this test?  Consider adding a comment to document your intent.

---

*[test/e2e/servicedns.go, line 106 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQnzWUCLC8FaX5cqSQ:-LEQnzWUCLC8FaX5cqSR:bpzay64) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L106)):*
> ```Go
> 				_, err := fedClient.MulticlusterdnsV1alpha1().MultiClusterServiceDNSRecords(namespace).Create(serviceDNS)
> 				framework.ExpectNoError(err, "Error creating MultiClusterServiceDNS object %v", serviceDNS)
> 				defer fedClient.MulticlusterdnsV1alpha1().MultiClusterServiceDNSRecords(namespace).Delete(serviceDNS.GetName(), nil)
> ```

Ditto avoiding the overhead of unnecessary resource deletion.

---

*[test/e2e/servicedns.go, line 115 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQpbrSBXCDZTB6i6XC:-LEQpbrT5na1LFEnejI7:b8s48sx) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L115)):*
> ```Go
> 				lbsuffix := 1
> 				for clusterName, client := range f.ClusterKubeClients(userAgent) {
> 					clusterLbs[clusterName] = fmt.Sprintf("%s.%d", lbPrefix, lbsuffix)
> ```

Why does the lb ip need to be stored in a map?  The only usage is in the same loop iteration.

---

*[test/e2e/framework/framework.go, line 53 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQhHj28THHfZFhU8fL:-LEQhHj28THHfZFhU8fM:b-a5rgzz) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/framework/framework.go#L53)):*
> ```Go
> 	SetUpControllerFixture(typeConfig typeconfig.Interface)
> 
> 	SetUpServiceDNSControllerFixture()
> ```

(No action required) Now that there is more than one feature, it probably makes sense to avoid putting feature-specific setup in the framework.

---

*[test/integration/servicedns_test.go, line 134 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQj_to6VaFana9ZsFq:-LEQj_to6VaFana9ZsFr:b-ks9ph8) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/integration/servicedns_test.go#L134)):*
> ```Go
> }
> 
> // serviceDNSControllerFixture manages the MultiClusterServiceDNS controller for testing.
> ```

Why isn't this in framework with the rest of the fixture?

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-:-LEQqQSY0Oz5-glxGCrV:b-jn247x)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
> What is the intended relationship between a MultiClusterServiceDNSRecord and a FederatedService when the 2 features are deployed together?  As proposed, there is no relation and a user would have to create and manage the 2 resources manually.  There is also no provision for targeting member clusters selectively.  I'm not suggesting that this question should block merge, but I would expect it be answered before alpha release if the goal is v1 parity.

MultiClusterServiceDNSRecord targets the FederatedService with the same name in the namespace. Also currently the behaviour is to select all the clusters in the federation and there is no explicit cluster selection to simplify user input.

> I'm also a bit confused at the duplication of testing effort. I won't block merge for this reason, but I would like to see removal of duplication going in future PRs.

Currently, both integration and e2e are testing similar stuff mostly. But most of the code reused among them. As @font suggested that we may remove integration test in favour of managed e2e tests in future.

---

Review status: 1 of 107 files reviewed, 16 unresolved discussions (waiting on @marun, @shashidharatd, and @font)

---

*[pkg/apis/federatedscheduling/v1alpha1/zz_generated.conversion.go, line 4 at r3](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQYgWUiubH5q92c9a6-r3-4:-LEjmnG37DZkQEjNzsG7:b-896fix) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/d59976f2b1f6d73bee40e9619f3edb6fda54bab7/pkg/apis/federatedscheduling/v1alpha1/zz_generated.conversion.go#L4)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Ah ok.  Good to know.
</blockquote></details>

Done.

---

*[pkg/apis/multiclusterdns/v1alpha1/multiclusterservicednsrecord_types.go, line 63 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQaL7vBZ-7Nupyrtrd:-LEjmpOm8GFcICObw4Ci:bgwwlu2) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/pkg/apis/multiclusterdns/v1alpha1/multiclusterservicednsrecord_types.go#L63)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why is there no json tag for this field?
</blockquote></details>

Missed. Added now.

---

*[pkg/controller/federatedcluster/clusterclient.go, line 114 at r2](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQYgWTR9kOQmem3XTi-r2-114:-LEjnkS-AE1eehuX8boV:bb7uc6m) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/807e7f96143c252705b6a3ed5baeb8e0f899c440/pkg/controller/federatedcluster/clusterclient.go#L114)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) Why is zones a slice if only a single value is ever returned?
</blockquote></details>

It was originally intended for multi-zone cluster, retaining the same api as was existing in federation v1. We can take call on this in future.

---

*[pkg/controller/servicedns/controller.go, line 369 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQcwsm5RBXxy4N2SmN:-LEjo7D1222Z8rkrcvz-:bg8210n) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/pkg/controller/servicedns/controller.go#L369)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Given that `endpoints` is only used to determine whether a service is backed by pods, why is it necessary to collect the addresses?
</blockquote></details>

changed the function signature to implement as suggested.

---

*[test/common/servicedns.go, line 64 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQfIJEDCp6i57RUIt4:-LEjp12n2agS0imjNDlv:b-l6t1kv) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/common/servicedns.go#L64)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) I'm curious why you chose to implement this from scratch rather than extending the functions in testobjects.go.

I'm also not sure why you're loading fixture from disk.  This was done for the push reconciler because the rewrite to the dynamic client made it easier to define test objects as yaml, because the equivalent definition with Unstructured are very verbose.  It also allows for user-defined primitives to be validated without writing code.  In the case of the `servicedns` controller, there is no need a generic approach, and defining fixture in code seems much simpler.
</blockquote></details>

Implemented back to define the objects in code instead of loading from disk.

---

*[test/common/servicedns.go, line 90 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQeC36EJUapr9bm2tf:-LEjoNxD4Nni8e6pLBtX:b-gjvn8z) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/common/servicedns.go#L90)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) This and the wait-for-deletion function seem like an awful lot of work for a single type.  There are already similar functions in crudtester.go that could be extended to accept functions for accessing and comparing a resource so that the type/client-specific behavior can be passed in rather than hard-coded.
</blockquote></details>

The utility function in crudtester.go are method recievers to FederatedTypeCrudTester and also work using dynamic clients. I felt it would be too much change.

---

*[test/common/servicedns.go, line 125 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQeWPF14AU-_wPaQyC:-LEjoLy7-CWKxUH5ia97:bk3gule) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/common/servicedns.go#L125)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

What about meta equivalency?
</blockquote></details>

updated now.

---

*[test/e2e/servicedns.go, line 62 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQl6E-DqM2qtEIu8pz:-LEjpPswCHsrxUYRWmWN:b-4tbw72) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L62)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider revising this test name to be more descriptive.  The test as written seems to validate that the servicedns controller is correctly maintaining state without a service or endpoints.
</blockquote></details>

Revised. But still not sure, whether its ideal.

---

*[test/e2e/servicedns.go, line 69 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQmHXhDmjGfVn9f1L3:-LEjrfaR4sw7aIJW76Jr:bz05qet) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L69)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why is it desirable to explicitly delete the resource, given that the namespace and its contents will be removed on framework teardown?
</blockquote></details>

This test is basically testing CRUD of ServiceDNS object, So the object is explicitly deleted.

---

*[test/e2e/servicedns.go, line 70 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQlFYuCGdJCmpQfdKT:-LEjpYzN0P0OKnYGxZq2:bkq4fpq) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L70)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why does the wait function accept a polling interval but not a timeout?  I think both should vary between integration and e2e as per the example of the crud tester.
</blockquote></details>

now accepting timeout too.

---

*[test/e2e/servicedns.go, line 90 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQlHfA9jhO_mJqnIF4:-LEjpb160pNp9rjEUrSH:b-qssebl) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L90)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Ditto the comment about varying the timeout.
</blockquote></details>

fixed.

---

*[test/e2e/servicedns.go, line 92 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQm3oX9Cd3aCvWfEmH:-LEjs07t24XZ5as_frx4:b-m07kqe) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L92)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

What is being checked in the remainder of this test?  Consider adding a comment to document your intent.
</blockquote></details>

Now added some By() clauses and comments.

---

*[test/e2e/servicedns.go, line 106 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQnzWUCLC8FaX5cqSQ:-LEjs8s72h0edKmNmSJh:b-1qg7aj) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L106)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Ditto avoiding the overhead of unnecessary resource deletion.
</blockquote></details>

ok. done

---

*[test/e2e/servicedns.go, line 115 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQpbrSBXCDZTB6i6XC:-LEjsEZG-ISvPrccNT1Q:b-cy40h3) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L115)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why does the lb ip need to be stored in a map?  The only usage is in the same loop iteration.
</blockquote></details>

ok fixed.

---

*[test/e2e/framework/framework.go, line 53 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQhHj28THHfZFhU8fL:-LEjsPBDEU_3I6bBNU7M:byvbloj) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/framework/framework.go#L53)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) Now that there is more than one feature, it probably makes sense to avoid putting feature-specific setup in the framework.
</blockquote></details>

ok will do refactor as suggested in future.

---

*[test/integration/servicedns_test.go, line 134 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQj_to6VaFana9ZsFq:-LEjsbB35XNdgoliupOE:b2eaayn) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/integration/servicedns_test.go#L134)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why isn't this in framework with the rest of the fixture?
</blockquote></details>

It does have some specific stuff related to the test case, like the setting up loadbalancer ip. It made sense to put them in same file.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-:-LEjt37X9ue43al6q6ri:btnlgw6)*
<!-- Sent from Reviewable.io -->

marun commented:
Reviewed 3 of 10 files at r7, 15 of 106 files at r8, 106 of 106 files at r9.
Review status: all files reviewed, 6 unresolved discussions (waiting on @marun and @shashidharatd)

---

*[pkg/controller/federatedcluster/clusterclient.go, line 114 at r2](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQYgWTR9kOQmem3XTi-r2-114:-LEpIT-_7oK6V0cpdX_0:b-asyoxs) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/807e7f96143c252705b6a3ed5baeb8e0f899c440/pkg/controller/federatedcluster/clusterclient.go#L114)):*
<details><summary><i>Previously, shashidharatd (shashidharatd) wrote…</i></summary><blockquote>

It was originally intended for multi-zone cluster, retaining the same api as was existing in federation v1. We can take call on this in future.
</blockquote></details>

(No action required) I think following YAGNI is best in cases like this - cut out what is not used.  It can always be added back if necessary.

---

*[pkg/controller/servicedns/controller.go, line 383 at r9](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEpJxvbCu2DGmuipPvR:-LEpJxvbCu2DGmuipPvS:b3peni0) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/373a3999eca6346c6ca56b8a7e561e563b3e6f5f/pkg/controller/servicedns/controller.go#L383)):*
> ```Go
> 	}
> 
> 	// We should preserve ClusterDNS entries (without loadbalancers) for offline clusters
> ```

I do not understand this comment.  What client is being referred to?  What is a 'service shard'?  How would the redirection take place?

---

*[pkg/controller/servicedns/controller.go, line 466 at r9](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEpIvL7-rwIK9uONvAv:-LEpIvL7-rwIK9uONvAw:b-mwqtuf) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/373a3999eca6346c6ca56b8a7e561e563b3e6f5f/pkg/controller/servicedns/controller.go#L466)):*
> ```Go
> 		return false, err
> 	}
> 	if endpointFound {
> ```

Will an endpoint ever have no addresses?  If no, consider returning here.  If yes, consider returning `true` on line 483. 

---

*[test/common/servicedns.go, line 90 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQeC36EJUapr9bm2tf:-LEpKzzYDjWVsDlYnWms:b-x3f134) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/common/servicedns.go#L90)):*
<details><summary><i>Previously, shashidharatd (shashidharatd) wrote…</i></summary><blockquote>

The utility function in crudtester.go are method recievers to FederatedTypeCrudTester and also work using dynamic clients. I felt it would be too much change.
</blockquote></details>

It would be easy enough to pass in a function that performed the creation.  Details like the client used or the name wouldn't have to leak to the helper.  

---

*[test/common/servicedns.go, line 125 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQeWPF14AU-_wPaQyC:-LEpLYuu3Ex8pLzLz8uR:b-7wix0k) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/common/servicedns.go#L125)):*
<details><summary><i>Previously, shashidharatd (shashidharatd) wrote…</i></summary><blockquote>

updated now.
</blockquote></details>

I think it would be better to reuse the existing [utility function](https://github.com/kubernetes-sigs/federation-v2/blob/master/pkg/controller/util/meta.go#L82).

---

*[test/e2e/servicedns.go, line 69 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQmHXhDmjGfVn9f1L3:-LEpLvYrA20xbIzVWYU0:b-k125f4) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L69)):*
<details><summary><i>Previously, shashidharatd (shashidharatd) wrote…</i></summary><blockquote>

This test is basically testing CRUD of ServiceDNS object, So the object is explicitly deleted.
</blockquote></details>

Testing of federated type deletion validates that cleanup from member clusters occurs.  If there are no such intended side effects here (e.g. finalizer-based cleanup), what is being testing here that suggests an e2e environment?

---

*[test/e2e/servicedns.go, line 92 at r6](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEQm3oX9Cd3aCvWfEmH:-LEpNKOXADMTSYPCfUsz:b-tgvep7) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/7b45126cf6f701d46583ca4a8d44f32b3c38fb6a/test/e2e/servicedns.go#L92)):*
<details><summary><i>Previously, shashidharatd (shashidharatd) wrote…</i></summary><blockquote>

Now added some By() clauses and comments.
</blockquote></details>

I apologize for not being clear.  What is the controller expected to do in response to the update?

---

*[test/e2e/servicedns.go, line 47 at r9](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEpO4Ja9_OkUnneeE7s:-LEpO4Ja9_OkUnneeE7t:b-bfz6u3) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/373a3999eca6346c6ca56b8a7e561e563b3e6f5f/test/e2e/servicedns.go#L47)):*
> ```Go
> 
> 	BeforeEach(func() {
> 		userAgent = "test-service-dns"
> ```

Why do these 2 static values need to vary for each test?  They seem like values that could be defined as constants.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59)*
<!-- Sent from Reviewable.io -->

font commented:
Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 89 of 90 files at r3, 3 of 6 files at r5, 5 of 11 files at r6, 10 of 10 files at r7.
Review status: all files reviewed, 7 unresolved discussions (waiting on @shashidharatd)

---

*[.gitignore, line 6 at r7](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LElfJsPBzc6v-k6R1Ag:-LElfJsPBzc6v-k6R1Ah:b-2maelm) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/6c30c16bfc4b2d0915683c6269535d77c80dfd0a/.gitignore#L6)):*
> ```
> /kubeconfig
> /config
> /.idea
> ```

I suggest specifying to ignore these files for your own local repo or globally for all your repos. This avoids everyone adding `<insert-your-favorite-editor>` specific ignores to the main repo.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59)*
<!-- Sent from Reviewable.io -->

marun commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 13 of 13 files at r10.
Review status: all files reviewed, 2 unresolved discussions (waiting on @font and @shashidharatd)

---

*[pkg/controller/federatedcluster/clusterclient.go, line 119 at r10](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEt_tUh-lJ-C0GWhf_d:-LEt_tUh-lJ-C0GWhf_e:b9jsfk7) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/fdca3fcea559da18988559997cb57f03817b85e0/pkg/controller/federatedcluster/clusterclient.go#L119)):*
> ```Go
> 		return "", "", err
> 	}
> 	for i, node := range nodes.Items {
> ```

(No action required) Could be simplified?

```golang
node := nodes.Items[0]
zone, err = getZoneNameNodeNode(node)
...
```

---

*[pkg/controller/servicedns/controller.go, line 383 at r9](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-LEpJxvbCu2DGmuipPvR:-LEtaaCI9_1MhacLyKrT:b-wtkc5h) ([raw file](https://github.com/kubernetes-sigs/federation-v2/blob/373a3999eca6346c6ca56b8a7e561e563b3e6f5f/pkg/controller/servicedns/controller.go#L383)):*
<details><summary><i>Previously, shashidharatd (shashidharatd) wrote…</i></summary><blockquote>

I rewrote the comment. PTAL. hope it is clear now.
</blockquote></details>

Yes, thank you.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kubernetes-sigs/federation-v2/59#-:-LEtbglt7tn7heLG4p_8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@font, can you PTAL and give a lgtm too. Thanks!

irfanurrehman commented:
@pmorie, this seems to be ok to merge now. Do you want to spend a cycle checking it, or should I go ahead and merge this?

pmorie commented:
I think this can be merged and further changes made in follow-up PRs.

irfanurrehman commented:
/lgtm
